lib/std/uniquelabels/unique_labels.go

```diff
@@ -0,0 +1,209 @@
+// Copyright (c) 2025 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package uniquelabels
+
+import (
+	"encoding/json"
+	"fmt"
+	"iter"
+	"maps"
+
+	"github.com/projectcalico/calico/lib/std/uniquestr"
+)
+
+type handleMap = map[uniquestr.Handle]uniquestr.Handle
+
+var (
+	Nil   = Map{m: nil}
+	Empty = Map{m: map[uniquestr.Handle]uniquestr.Handle{}}
+)
+
+// Map is a *read only* string-to-string map that interns keys and values.
+// When marshalled as JSON, uses the same representation as a normal
+// map[string]string.  When unmarshalled, keys and values get interned/deduped.
+//
+// To allow drop-in replacement of a map[string]string, Map has unique
+// representations of nil (Nil) and empty (Empty). The zero value of a Map is
+// the Nil value; it can be detected with IsNil(). Map is a value type, which
+// precludes using Go's normal nil.  Making Map a pointer type would increase
+// overheads; making map an alias for map[uniquestr.Handle]uniquestr.Handle
+// would break the read-only property, preventing future optimisations.
+//
+// Uses uniquestr.Handle internally, so it is most efficient to query the map
+// using AllHandles() and GetHandle().
+type Map struct {
```

Comment:
Not really feedback, but for my own understanding I am curious if you considered a single `unique` package with `type Labels` and `type String` instead of separate `uniquelabels` and `uniquestring` packages and what led you in this direction.

Reply:
I tried making `uniquestring` just `unique` but the ambiguity with the stdlib package of the same name was a bit confusing.  I tried a few variations with no super clear winner TBH!  Happy to get a second opinion.